### PR TITLE
feat(llma): add evaluation results API endpoint

### PR DESCRIPTION
**Pull request by andrewm4894**
## Problem

Evaluation results (`$ai_evaluation` events) are stored in ClickHouse but there's no REST API endpoint to query them. The frontend uses inline HogQL queries, and the MCP server would need to embed raw HogQL too. This is fragile — if the event schema changes, every consumer breaks.

## Changes

Add `GET /api/environments/:id/llm_analytics/evaluation_results/` endpoint that queries `$ai_evaluation` events from ClickHouse using HogQL AST (safe parameterization).

**Query params:**
- `evaluation_id` — filter by evaluation UUID
- `generation_id` — filter by generation event UUID
- `result` — filter by status: `pass`, `fail`, or `na`
- `limit` — max results (default 50, max 200)

At least one of `evaluation_id` or `generation_id` is required.

**Returns:** `{ results: [...], count: N }` with each result containing:
- `uuid`, `timestamp`, `evaluation_id`, `evaluation_name`
- `generation_id`, `trace_id`
- `result` (bool/null), `reasoning`, `applicable`

Includes unit tests with mocked ClickHouse queries.

**Part 2 of a 3-PR stack** — depends on PR #50004, MCP tool follows in PR #50011.

## How did you test this code?

- Unit tests in `test_evaluation_results.py` (mocked ClickHouse)
- Follows existing patterns from `evaluation_summary.py`
- Uses HogQL AST for safe query construction (no f-strings with user data)

Closes #50025 (tracked)

### CODE REVIEW

File: products/llm_analytics/backend/api/evaluation_results.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    permission_classes = [IsAuthenticated, AccessControlPermission]
+    serializer_class = EvaluationResultSerializer
+
+    @llma_track_latency("llma_evaluation_results_list")
+    @monitor(feature=None, endpoint="llma_evaluation_results_list", method="GET")
+    def list(self, request: Request, **kwargs) -> Response:
+        evaluation_id = request.query_params.get("evaluation_id")
+        generation_id = request.query_params.get("generation_id")
+        result_filter = request.query_params.get("result")
+        limit = min(int(request.query_params.get("limit", DEFAULT_LIMIT)), MAX_LIMIT)
```
[the remark this answers is not in the document]
Comment (andrewm4894) [reply products/llm_analytics/backend/api/evaluation_results.py]: Fixed — wrapped in try/except, returns 400 on invalid input. Added test coverage.

[the remark this answers is not in the document]
Comment (andrewm4894) [reply products/llm_analytics/backend/api/evaluation_results.py]: Fixed — same as Codex comment above, wrapped in try/except.

File: products/llm_analytics/backend/api/evaluation_results.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+                    left=ast.Field(chain=["properties", "$ai_target_event_id"]),
+                    right=ast.Constant(value=generation_id),
+                )
+            )
+
+        if result_filter == "pass":
+            where_conditions.append(
+                ast.CompareOperation(
+                    op=ast.CompareOperationOp.Eq,
+                    left=ast.Field(chain=["properties", "$ai_evaluation_result"]),
```
[the remark this answers is not in the document]
Comment (andrewm4894) [reply products/llm_analytics/backend/api/evaluation_results.py:106]: Good catch — added a `not_na_guard` (isNull OR != false) to both pass and fail filters, matching the pattern in `evaluation_summary.py`.

File: products/llm_analytics/backend/api/evaluation_results.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+                )
+            )
+        elif result_filter == "na":
+            where_conditions.append(
+                ast.CompareOperation(
+                    op=ast.CompareOperationOp.Eq,
+                    left=ast.Field(chain=["properties", "$ai_evaluation_applicable"]),
+                    right=ast.Constant(value=False),
+                )
+            )
```
[the remark this answers is not in the document]
Comment (andrewm4894) [reply products/llm_analytics/backend/api/evaluation_results.py:127]: Fixed — added the applicable guard for both pass and fail filters.

File: products/llm_analytics/backend/api/evaluation_results.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+                placeholders={
+                    "where_clause": ast.And(exprs=where_conditions),
+                    "limit": ast.Constant(value=limit),
+                },
+                team=self.team,
+            )
+
+        results = [
+            {
+                "uuid": str(row[0]) if row[0] else "",
```
[the remark this answers is not in the document]
Comment (andrewm4894) [reply products/llm_analytics/backend/api/evaluation_results.py]: Fair point — changed to `str(row[0])` directly. UUID from events table should never be null.

File: products/llm_analytics/backend/api/evaluation_results.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+    scope_object = "evaluation"
+    permission_classes = [IsAuthenticated, AccessControlPermission]
+    serializer_class = EvaluationResultSerializer
+
+    @llma_track_latency("llma_evaluation_results_list")
+    @monitor(feature=None, endpoint="llma_evaluation_results_list", method="GET")
+    def list(self, request: Request, **kwargs) -> Response:
+        evaluation_id = request.query_params.get("evaluation_id")
+        generation_id = request.query_params.get("generation_id")
+        result_filter = request.query_params.get("result")
```
Comment (sortafreel) [inline products/llm_analytics/backend/api/evaluation_results.py:53]: ### Unknown `result` values silently bypass filtering

**Priority:** should_fix | **Category:** bug | **Lines:** 53, 102-127

---

`result` is documented as `pass|fail|na`, but any other value falls through the conditional chain and returns unfiltered rows for the provided identifiers. A typo like `?result=pas` therefore returns logically incorrect data instead of a client error, which can hide caller bugs.

<details>
<summary><strong>Suggested fix</strong></summary>
<br>

Validate `result` before building `where_conditions` (e.g., allowlist `{"pass","fail","na"}`) and return HTTP 400 for unknown values. Keep the current branch logic only for validated values.

</details>

<details>
<summary><strong>Why we think it's a valid issue</strong></summary>
<br>

In `EvaluationResultsViewSet.list`, `result_filter` is taken from query params and only mapped for `pass`, `fail`, and `na` (around lines 102-127). Any other provided value falls through and no result-specific predicate is added, so the endpoint returns rows filtered only by `evaluation_id`/`generation_id`. That is a silent semantic failure: callers asking for a specific result can receive mixed results without an error, which hides client bugs and makes debugging harder. Given this endpoint already returns `400` for malformed `limit` and missing required identifiers, rejecting unknown `result` values with `400` is consistent and should be addressed.

</details>

<details>
<summary><strong>Prompt to fix with AI (copy-paste)</strong></summary>
<br>

```
## Context
@products/llm_analytics/backend/api/evaluation_results.py#L53
@products/llm_analytics/backend/api/evaluation_results.py#L102-127

<issue_description>
`result` is documented as `pass|fail|na`, but any other value falls through the conditional chain and returns unfiltered rows for the provided identifiers. A typo like `?result=pas` therefore returns logically incorrect data instead of a client error, which can hide caller bugs.
</issue_description>

<issue_validation>
In `EvaluationResultsViewSet.list`, `result_filter` is taken from query params and only mapped for `pass`, `fail`, and `na` (around lines 102-127). Any other provided value falls through and no result-specific predicate is added, so the endpoint returns rows filtered only by `evaluation_id`/`generation_id`. That is a silent semantic failure: callers asking for a specific result can receive mixed results without an error, which hides client bugs and makes debugging harder. Given this endpoint already returns `400` for malformed `limit` and missing required identifiers, rejecting unknown `result` values with `400` is consistent and should be addressed.
</issue_validation>

## Task
Investigate the issue and solve it

<potential_solution>
Validate `result` before building `where_conditions` (e.g., allowlist `{"pass","fail","na"}`) and return HTTP 400 for unknown values. Keep the current branch logic only for validated values.
</potential_solution>
```

</details>

File: products/llm_analytics/backend/api/evaluation_results.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+    GET /api/environments/:id/llm_analytics/evaluation_results/
+        ?evaluation_id=<uuid>       Filter by evaluation UUID
+        &generation_id=<uuid>       Filter by generation event UUID
+        &result=pass|fail|na        Filter by result status
+        &limit=50                   Max results (default 50, max 200)
+    """
+
+    scope_object = "evaluation"
+    permission_classes = [IsAuthenticated, AccessControlPermission]
```
Comment (sortafreel) [inline products/llm_analytics/backend/api/evaluation_results.py:45]: ### List endpoint can bypass evaluation object-level RBAC

**Priority:** must_fix | **Category:** security | **Lines:** 44-45, 73-89, 149-157

---

`scope_object` is set to `"evaluation"`, so `AccessControlPermission` permits `list` when a user has specific access to any evaluation object. This custom list implementation then executes a raw HogQL query without constraining results to evaluations the caller can access. A user with specific access to evaluation A but no resource-level evaluation access can still query by `generation_id` (or another `evaluation_id`) and retrieve results for evaluations they are not authorized to read.

<details>
<summary><strong>Suggested fix</strong></summary>
<br>

Enforce object-level filtering before executing HogQL. Build an allowlist of accessible evaluation IDs (for example, via `self.user_access_control.filter_queryset_by_access_level(Evaluation.objects.filter(team_id=self.team_id))`) and add an `IN` predicate on `properties.$ai_evaluation_id` to `where_conditions`; if no IDs are accessible, return an empty result (or 403). Add an RBAC test case where resource-level `evaluation` access is `none` but specific access exists for one evaluation, and verify the endpoint cannot return rows for other evaluations.

</details>

<details>
<summary><strong>Why we think it's a valid issue</strong></summary>
<br>

The issue is valid. In `EvaluationResultsViewSet` (`products/llm_analytics/backend/api/evaluation_results.py`), `scope_object` is `"evaluation"`, and `AccessControlPermission.has_permission` explicitly allows `list` when a user has any specific object grant for that resource (even if resource-level access is `none`). Normally list safety then depends on queryset filtering (`filter_queryset_by_access_level`), but this endpoint uses a custom `list` method with a raw HogQL query and never applies object-level filtering against accessible `Evaluation` IDs. The query only constrains by team/event and optional user-supplied `evaluation_id`/`generation_id`, so a user with access to evaluation A can retrieve rows for evaluation B in the same team if they supply matching identifiers. That is an RBAC data-exposure gap and should be fixed with access-constrained filtering plus a regression test for specific-access-only users.

</details>

<details>
<summary><strong>Prompt to fix with AI (copy-paste)</strong></summary>
<br>

```
## Context
@products/llm_analytics/backend/api/evaluation_results.py#L44-45
@products/llm_analytics/backend/api/evaluation_results.py#L73-89
@products/llm_analytics/backend/api/evaluation_results.py#L149-157

<issue_description>
`scope_object` is set to `"evaluation"`, so `AccessControlPermission` permits `list` when a user has specific access to any evaluation object. This custom list implementation then executes a raw HogQL query without constraining results to evaluations the caller can access. A user with specific access to evaluation A but no resource-level evaluation access can still query by `generation_id` (or another `evaluation_id`) and retrieve results for evaluations they are not authorized to read.
</issue_description>

<issue_validation>
The issue is valid. In `EvaluationResultsViewSet` (`products/llm_analytics/backend/api/evaluation_results.py`), `scope_object` is `"evaluation"`, and `AccessControlPermission.has_permission` explicitly allows `list` when a user has any specific object grant for that resource (even if resource-level access is `none`). Normally list safety then depends on queryset filtering (`filter_queryset_by_access_level`), but this endpoint uses a custom `list` method with a raw HogQL query and never applies object-level filtering against accessible `Evaluation` IDs. The query only constrains by team/event and optional user-supplied `evaluation_id`/`generation_id`, so a user with access to evaluation A can retrieve rows for evaluation B in the same team if they supply matching identifiers. That is an RBAC data-exposure gap and should be fixed with access-constrained filtering plus a regression test for specific-access-only users.
</issue_validation>

## Task
Investigate the issue and solve it

<potential_solution>
Enforce object-level filtering before executing HogQL. Build an allowlist of accessible evaluation IDs (for example, via `self.user_access_control.filter_queryset_by_access_level(Evaluation.objects.filter(team_id=self.team_id))`) and add an `IN` predicate on `properties.$ai_evaluation_id` to `where_conditions`; if no IDs are accessible, return an empty result (or 403). Add an RBAC test case where resource-level `evaluation` access is `none` but specific access exists for one evaluation, and verify the endpoint cannot return rows for other evaluations.
</potential_solution>
```

</details>

File: products/llm_analytics/backend/api/evaluation_results.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+            {
+                "uuid": str(row[0]),
+                "timestamp": row[1],
+                "evaluation_id": str(row[2]) if row[2] else "",
+                "evaluation_name": row[3] or "",
+                "generation_id": str(row[4]) if row[4] else "",
+                "trace_id": str(row[5]) if row[5] else None,
+                "result": None if row[8] is False else row[6],
+                "reasoning": row[7] or None,
+                "applicable": row[8],
```
Comment (sortafreel) [inline products/llm_analytics/backend/api/evaluation_results.py:169]: ### Boolean-like HogQL values are returned without normalization

**Priority:** should_fix | **Category:** bug | **Lines:** 167-169

---

The response mapping assumes `row[8]` is a native Python boolean (`row[8] is False`) and forwards `row[6]`/`row[8]` directly. For event properties read through HogQL/property-group columns, boolean-like values can arrive as strings (`"true"`/`"false"`) before typed materialization is guaranteed. In that case, the N/A normalization branch is skipped and the endpoint can emit string values for `result`/`applicable`, violating the declared response contract and causing unreliable client behavior.

<details>
<summary><strong>Suggested fix</strong></summary>
<br>

Normalize `result` and `applicable` to `bool | None` before building each response row, then compute `result = None if applicable is False else normalized_result`. Add a helper that maps `{True, False, None, 'true', 'false', 1, 0}` to canonical booleans/null, and add a test case with mocked string booleans to lock the behavior.

</details>

<details>
<summary><strong>Why we think it's a valid issue</strong></summary>
<br>

The issue is valid. In `evaluation_results.py`, the row mapping assumes `applicable` is a native `False` object (`row[8] is False`) and otherwise forwards `row[6]`/`row[8]` directly. For HogQL property access (`properties.$...`), values are often string-backed unless property typing/materialization has been resolved for that team/property, and this is already reflected in existing llm analytics frontend code that explicitly handles `'true'`/`'false'`. Because this view returns raw dicts (it does not serialize output through `EvaluationResultSerializer`), string/int boolean-like values can be emitted directly, violating the endpoint contract (`bool | null`) and causing incorrect N/A handling (e.g. `'false'` does not trigger `result = None`). Risk: contract/type mismatch for API consumers (including MCP/typed clients), inconsistent client behavior, and incorrect result interpretation.

</details>

<details>
<summary><strong>Prompt to fix with AI (copy-paste)</strong></summary>
<br>

```
## Context
@products/llm_analytics/backend/api/evaluation_results.py#L167-169

<issue_description>
The response mapping assumes `row[8]` is a native Python boolean (`row[8] is False`) and forwards `row[6]`/`row[8]` directly. For event properties read through HogQL/property-group columns, boolean-like values can arrive as strings (`"true"`/`"false"`) before typed materialization is guaranteed. In that case, the N/A normalization branch is skipped and the endpoint can emit string values for `result`/`applicable`, violating the declared response contract and causing unreliable client behavior.
</issue_description>

<issue_validation>
The issue is valid. In `evaluation_results.py`, the row mapping assumes `applicable` is a native `False` object (`row[8] is False`) and otherwise forwards `row[6]`/`row[8]` directly. For HogQL property access (`properties.$...`), values are often string-backed unless property typing/materialization has been resolved for that team/property, and this is already reflected in existing llm analytics frontend code that explicitly handles `'true'`/`'false'`. Because this view returns raw dicts (it does not serialize output through `EvaluationResultSerializer`), string/int boolean-like values can be emitted directly, violating the endpoint contract (`bool | null`) and causing incorrect N/A handling (e.g. `'false'` does not trigger `result = None`). Risk: contract/type mismatch for API consumers (including MCP/typed clients), inconsistent client behavior, and incorrect result interpretation.
</issue_validation>

## Task
Investigate the issue and solve it

<potential_solution>
Normalize `result` and `applicable` to `bool | None` before building each response row, then compute `result = None if applicable is False else normalized_result`. Add a helper that maps `{True, False, None, 'true', 'false', 1, 0}` to canonical booleans/null, and add a test case with mocked string booleans to lock the behavior.
</potential_solution>
```

</details>

File: products/llm_analytics/frontend/generated/api.schemas.ts [hunk 8]
```diff
@@ -797,6 +822,17 @@ export type LlmAnalyticsClusteringJobsListParams = {
...
+    /**
+     * Number of results to return per page.
+     */
+    limit?: number
+    /**
+     * The initial index from which to return the results.
+     */
+    offset?: number
+}
+
```
Comment (sortafreel) [inline products/llm_analytics/frontend/generated/api.schemas.ts:835]: ### Generated params type omits required/supported filters for evaluation results

**Priority:** must_fix | **Category:** bug | **Lines:** 825-835

---

`LlmAnalyticsEvaluationResultsListParams` only includes `limit` and `offset`, but the backend endpoint requires at least one of `evaluation_id` or `generation_id` and also supports `result`. In TypeScript consumers, passing these filter fields is a type error, so the generated helper is not usable for valid requests without unsafe casts.

<details>
<summary><strong>Suggested fix</strong></summary>
<br>

Add explicit OpenAPI query parameter annotations on `EvaluationResultsViewSet.list` for `evaluation_id`, `generation_id`, and `result`, then regenerate OpenAPI artifacts. This should produce a params type that lets typed callers send valid filters.

</details>

<details>
<summary><strong>Why we think it's a valid issue</strong></summary>
<br>

This issue is valid. In `products/llm_analytics/frontend/generated/api.schemas.ts`, `LlmAnalyticsEvaluationResultsListParams` only includes `limit` and `offset`, but the backend endpoint in `products/llm_analytics/backend/api/evaluation_results.py` reads `evaluation_id`, `generation_id`, and `result`, and explicitly returns `400` if neither ID is provided. As generated, typed callers cannot pass required/supported filters without unsafe casts, and they can compile requests that fail at runtime. This indicates missing OpenAPI query parameter declarations for the endpoint; adding explicit parameter annotations and regenerating artifacts is the correct fix. Leaving this as-is creates a type/API contract mismatch, forcing `any` usage and increasing runtime error risk.

</details>

<details>
<summary><strong>Prompt to fix with AI (copy-paste)</strong></summary>
<br>

```
## Context
@products/llm_analytics/frontend/generated/api.schemas.ts#L825-835

<issue_description>
`LlmAnalyticsEvaluationResultsListParams` only includes `limit` and `offset`, but the backend endpoint requires at least one of `evaluation_id` or `generation_id` and also supports `result`. In TypeScript consumers, passing these filter fields is a type error, so the generated helper is not usable for valid requests without unsafe casts.
</issue_description>

<issue_validation>
This issue is valid. In `products/llm_analytics/frontend/generated/api.schemas.ts`, `LlmAnalyticsEvaluationResultsListParams` only includes `limit` and `offset`, but the backend endpoint in `products/llm_analytics/backend/api/evaluation_results.py` reads `evaluation_id`, `generation_id`, and `result`, and explicitly returns `400` if neither ID is provided. As generated, typed callers cannot pass required/supported filters without unsafe casts, and they can compile requests that fail at runtime. This indicates missing OpenAPI query parameter declarations for the endpoint; adding explicit parameter annotations and regenerating artifacts is the correct fix. Leaving this as-is creates a type/API contract mismatch, forcing `any` usage and increasing runtime error risk.
</issue_validation>

## Task
Investigate the issue and solve it

<potential_solution>
Add explicit OpenAPI query parameter annotations on `EvaluationResultsViewSet.list` for `evaluation_id`, `generation_id`, and `result`, then regenerate OpenAPI artifacts. This should produce a params type that lets typed callers send valid filters.
</potential_solution>
```

</details>

File: products/llm_analytics/frontend/generated/api.ts [hunk 9]
```diff
@@ -355,6 +357,45 @@ export const llmAnalyticsEvaluationConfigSetActiveKeyCreate = async (
...
+        if (value !== undefined) {
+            normalizedParams.append(key, value === null ? 'null' : value.toString())
+        }
+    })
+
+    const stringifiedParams = normalizedParams.toString()
+
+    return stringifiedParams.length > 0
+        ? `/api/environments/${projectId}/llm_analytics/evaluation_results/?${stringifiedParams}`
+        : `/api/environments/${projectId}/llm_analytics/evaluation_results/`
```
Comment (sortafreel) [inline products/llm_analytics/frontend/generated/api.ts:385]: ### Client exposes `offset` pagination parameter that the endpoint does not apply

**Priority:** should_fix | **Category:** bug | **Lines:** 369-385

---

The URL builder forwards all params (including `offset`), but the backend query logic currently applies only `LIMIT` and no offset. Callers attempting pagination will repeatedly fetch the first slice, leading to incorrect behavior.

<details>
<summary><strong>Suggested fix</strong></summary>
<br>

Align behavior and contract: either implement `offset` support in the backend query (with consistent pagination semantics) or remove `offset` from the endpoint schema so generated clients cannot send a misleading no-op parameter. Regenerate the generated frontend files after fixing the backend schema/behavior.

</details>

<details>
<summary><strong>Why we think it's a valid issue</strong></summary>
<br>

The issue is valid. In the generated client, `LlmAnalyticsEvaluationResultsListParams` includes `offset` and the URL helper forwards it (`Object.entries(params)`), but the backend endpoint (`products/llm_analytics/backend/api/evaluation_results.py`) only parses `limit` and builds a query with `LIMIT {limit}` and no offset clause. That means requests with different offsets return the same first page, so pagination behavior is incorrect and misleading for consumers. This contract mismatch can cause duplicate data in paginated UIs, broken infinite-scroll logic, and unnecessary repeated API calls. The API schema and backend behavior should be aligned (either implement offset semantics server-side or remove offset from the schema/client surface).

</details>

<details>
<summary><strong>Prompt to fix with AI (copy-paste)</strong></summary>
<br>

```
## Context
@products/llm_analytics/frontend/generated/api.ts#L369-385

<issue_description>
The URL builder forwards all params (including `offset`), but the backend query logic currently applies only `LIMIT` and no offset. Callers attempting pagination will repeatedly fetch the first slice, leading to incorrect behavior.
</issue_description>

<issue_validation>
The issue is valid. In the generated client, `LlmAnalyticsEvaluationResultsListParams` includes `offset` and the URL helper forwards it (`Object.entries(params)`), but the backend endpoint (`products/llm_analytics/backend/api/evaluation_results.py`) only parses `limit` and builds a query with `LIMIT {limit}` and no offset clause. That means requests with different offsets return the same first page, so pagination behavior is incorrect and misleading for consumers. This contract mismatch can cause duplicate data in paginated UIs, broken infinite-scroll logic, and unnecessary repeated API calls. The API schema and backend behavior should be aligned (either implement offset semantics server-side or remove offset from the schema/client surface).
</issue_validation>

## Task
Investigate the issue and solve it

<potential_solution>
Align behavior and contract: either implement `offset` support in the backend query (with consistent pagination semantics) or remove `offset` from the endpoint schema so generated clients cannot send a misleading no-op parameter. Regenerate the generated frontend files after fixing the backend schema/behavior.
</potential_solution>
```

</details>

**Review by sortafreel — COMMENTED** [review]
# ReviewHog Report

## Api

**Files:** `products/llm_analytics/backend/api/evaluation_results.py`, `products/llm_analytics/backend/api/__init__.py`, `posthog/api/__init__.py`

**Issues:** 3 issues

<details>
<summary>What were the main changes</summary>
<br>

- Introduces a new `EvaluationResultsViewSet` endpoint to query `$ai_evaluation` events with filters for `evaluation_id`, `generation_id`, `result`, and `limit`.
- Builds ClickHouse conditions with HogQL AST placeholders for safe parameterization, including the `not_na_guard` so `pass`/`fail` filters exclude N/A rows.
- Adds request validation and error handling in the endpoint flow (required identifier guard and invalid `limit` handling returning 400).
- Wires the viewset into llm analytics API exports and registers `/api/environments/:id/llm_analytics/evaluation_results/` on the environments router.

</details>

<details>
<summary>Full analysis</summary>
<br>

This chunk adds a dedicated REST read path for LLM evaluation event data: `GET /api/environments/:id/llm_analytics/evaluation_results/`. Its core goal is to move consumers away from embedding raw HogQL in frontend/MCP callers and centralize `$ai_evaluation` retrieval behind a stable backend contract. The new `EvaluationResultsViewSet` queries ClickHouse event rows, supports server-side filters (`evaluation_id`, `generation_id`, `result`, `limit`), enforces that at least one identifier is provided to avoid broad scans, and normalizes output fields (`uuid`, IDs, reasoning, applicability, and null-result semantics for N/A rows).

Architecturally, the endpoint follows existing LLM analytics API patterns: DRF `GenericViewSet` + `TeamAndOrgViewSetMixin`, standard auth/RBAC scope (`scope_object="evaluation"`), and observability decorators (`llma_track_latency`, `@monitor`) plus ClickHouse query tagging (`tags_context(product=Product.LLM_ANALYTICS)`). Query construction uses HogQL AST expressions and placeholders (`ast.CompareOperation`, `ast.And`, `ast.Constant`) instead of string interpolation, matching the secure pattern used in `evaluation_summary.py`. The `not_na_guard` added to pass/fail filtering aligns read behavior with how `$ai_evaluation_applicable` is emitted in the evaluation workflow and prevents pass/fail responses from leaking N/A events.

Integration-wise, this chunk wires the endpoint into both module exports and the global environments router (`products/llm_analytics/backend/api/__init__.py` and `posthog/api/__init__.py`), making it part of the product API surface for downstream clients. Tests around this endpoint mock `execute_hogql_query` and validate key API behavior (auth, required identifier guard, N/A mapping, max-limit capping, invalid-limit 400). Reviewer-raised concerns were incorporated in this chunk (limit parse error handling, pass/fail N/A exclusion, fail-fast UUID mapping), while parameter validation remains intentionally lightweight/manual rather than serializer-driven for query params.

</details>

## Frontend

**Files:** `products/llm_analytics/frontend/generated/api.schemas.ts`, `products/llm_analytics/frontend/generated/api.ts`

**Issues:** 2 issues

<details>
<summary>What were the main changes</summary>
<br>

- Regenerates frontend API schema types to include evaluation result entities and list response/params types for the new endpoint.
- Adds generated frontend API client helpers (`getLlmAnalyticsEvaluationResultsListUrl` and `llmAnalyticsEvaluationResultsList`) for calling the endpoint.
- Aligns generated frontend integration surface with the newly exposed backend route so consumers do not need inline HogQL.

</details>

<details>
<summary>Full analysis</summary>
<br>

This chunk regenerates the LLM analytics frontend OpenAPI artifacts so the new evaluation-results backend endpoint has first-class TypeScript contracts. In `products/llm_analytics/frontend/generated/api.schemas.ts`, it adds `EvaluationResultApi` (UUID/timestamp/evaluation identifiers plus nullable `trace_id`, `result`, `reasoning`, and `applicable`) and `PaginatedEvaluationResultListApi`, along with `LlmAnalyticsEvaluationResultsListParams`. The schema shape mirrors the backend serializer and keeps response fields read-only, signaling that these records are event-derived and server-owned.

In `products/llm_analytics/frontend/generated/api.ts`, the generated import surface is extended and a standard Orval pair is added: `getLlmAnalyticsEvaluationResultsListUrl` for URL/query-string construction and `llmAnalyticsEvaluationResultsList` for issuing the GET request. The function follows existing module conventions (generic `URLSearchParams` normalization and `apiMutator` transport), so it plugs into PostHog’s shared HTTP stack (`api-orval-mutator` -> `lib/api`) without introducing a new client abstraction. This keeps integration consistent with all other generated LLM analytics endpoints.

Architecturally, this chunk is an integration bridge between DRF/drf-spectacular backend contracts and typed frontend consumers, and it prepares later layers (including MCP-oriented consumers in the PR stack) to stop relying on inline HogQL tuple parsing. A key technical insight is that the generated params type currently includes only pagination keys (`limit`, `offset`), while the backend endpoint also supports `evaluation_id`, `generation_id`, and `result` filters parsed manually in the viewset; this indicates the OpenAPI surface is partially inferred and may need explicit backend schema annotations if strict typed clients are expected to use the full filter contract.

</details>
